### DISCUSSION TITLE
How do you send data to a promise invocation?

### DISCUSSION THREAD

**Discussion by mattpeskett** [statelyai/xstate#1329] [General] [answered]
I'm using services that return promises extensively in my state machine. I'd like to send arguments to those services to perform some basic routing logic but I can't see a way to send arguments to those promises. Is there a way I can achieve this?

**Comment by davidkpiano** [comment, marked as answer]
You can use `data` to do that:

```js
// ...
invoke: {
  src: 'fetchUser',
  data: { id: (ctx) => ctx.userId }
},

// ...

services: {
  fetchUser: (ctx, event, { data }) => {
    return fetchUserFromAPI(data.id);
  }
}
```

Or do you mean something else? Can you show in code what you want to do?

**Reply by mattpeskett** [reply]
That's exactly what I mean, however, when I implement that I don't get a third parameter in the 'fetchUser' function. `ctx` and `event` are as expected but the third parameter is undefined and as such if I put `{ data }` I get an exception saying I cannot destructure property of undefined or null.

**Reply by davidkpiano** [reply]
What version of XState are you using? Ensure you're using the latest version.

**Reply by mattpeskett** [reply]
I was using 4.10.0 but I just tried with 4.11.0 and got the same result. In the [documentation](https://xstate.js.org/docs/guides/communication.html#the-invoke-property) it says that data is only used when invoking machines. I'm invoking a promise, so should I expect the data property to be available?

**Reply by mattpeskett** [reply]
Actually, I got a bit of a different result with 4.11.0. The third parameter to the fetchUser function isn't undefined, its 
```
{
   data: "undefined"
}
```

So still not what I expected but I'm at least getting a third parameter now.

**Reply by davidkpiano** [reply]
Can you share your full code via CodeSandbox?

**Reply by mattpeskett** [reply]
I haven't used code sandbox before so I didn't get it running but I put the basics of what I'm doing here https://codesandbox.io/s/youthful-hamilton-k09ph?file=/src/index.js. Thanks for the ongoing help btw!

**Reply by Andarist** [reply]
I couldn't run your codesandbox for some reason (I expect the problem on their side) but I've copy-pasted your code (and fixed it a little bit - you might have forgot to save some changes) and it seems to work OK. Check this out: https://codesandbox.io/s/dank-water-s27ku?file=/src/index.js

**Reply by mattpeskett** [reply]
Thanks @davidkpiano and @Andarist. My issue seems to have disappeared and it works exactly as you've explained. Thanks for your help.
